File: Graph/Java/GraphTutorial/src/Representation/WeightedGraph.java
package Representation;

import java.util.*;
import javafx.util.Pair;

public class WeightedGraph {

    private static LinkedList<Pair<Integer, Integer>>[] ConstructGraph(int V) {
        LinkedList<Pair<Integer, Integer>> adjListArray[] = new LinkedList[V];
        for (int i = 0; i < V; i++) {
            // every index in the adjListArray contains an LinkedList 
            // for the adjacent vertices 
            adjListArray[i] = new LinkedList<>();
        }
        return adjListArray;
    }

    private static void addEdge(LinkedList<Pair<Integer, Integer>>[] adjListArray, int src, int dest, int weight) {
        // Add an edge from src to dest
        adjListArray[src].add(new Pair<>(dest, weight)); // src --> dest
        // Since graph is undirected, add an edge from dest to src also 
        adjListArray[dest].add(new Pair<>(src, weight)); // dest --> src 
    }

    private static void printGraph(LinkedList<Pair<Integer, Integer>>[] graph) {
        for (int vertex = 0; vertex < graph.length; vertex++) {
            System.out.println("Node " + vertex + " makes an edge with ");
            for (Pair<Integer, Integer> pair : graph[vertex]) {
                System.out.println("\tNode" + pair.getKey() + " with edge weight= " + pair.getValue());
            }
            System.out.println("\n");
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int V = sc.nextInt(); // number of vertices
        LinkedList<Pair<Integer, Integer>>[] graph = ConstructGraph(V);
        addEdge(graph, 0, 1, 10);
        addEdge(graph, 0, 4, 20);
        addEdge(graph, 1, 2, 30);
        addEdge(graph, 1, 3, 40);
        addEdge(graph, 1, 4, 50);
        addEdge(graph, 2, 3, 60);
        addEdge(graph, 3, 4, 70);
        // print the adjacency list representation of the above graph
        printGraph(graph);
    }
}

File: Graph/CPP/Representation/WeightedGraph.cpp
#include <bits/stdc++.h>
using namespace std;

void addEdge(vector<pair<int, int>> adj[], int u, int v, int wt)
{
  adj[u].push_back(make_pair(v, wt)); // u -> v with cost wt
  adj[v].push_back(make_pair(u, wt)); // u -> v with cost wt (for un-directed graph)
}
void printGraph(vector<pair<int, int>> adj[], int V)
{
  for (int v = 0; v < V; ++v)
  {
    cout << "\n Node " << v << " Makes an Edge with:\n";
    for (auto x : adj[v])
    {
      int node = x.first;
      int weight = x.second;
      cout << "\tNode " << node << " with cost= " << weight << endl;
    }
    cout << "\n";
  }
}
int main()
{
  int V = 5;
  vector<pair<int, int>> adj[V];
  addEdge(adj, 0, 1, 10);
  addEdge(adj, 0, 4, 20);
  addEdge(adj, 1, 2, 30);
  addEdge(adj, 1, 3, 40);
  addEdge(adj, 1, 4, 50);
  addEdge(adj, 2, 3, 60);
  addEdge(adj, 3, 4, 70);
  printGraph(adj, V);
  return 0;
} 

/*
// ============================================
// ================ User Input ================
// ============================================
int main()
{
  int V, E;
  cout << "Num of Vertices: ";
  cin >> V;
  cout << "Num of Edges: ";
  cin >> E;
  vector<pair<int, int>> adj[V];

  int a, b, w;
  for (int i = 0; i < E; i++)
  {
    cout << "Enter a ,b and weight:" << endl;
    cin >> a >> b >> w;
    addEdge(adj, a, b, w);
  }
  printGraph(adj, V);
  return 0;
}
*/
